src/poetry/core/packages/constraints/any_constraint.py

```diff
@@ -1,33 +1,27 @@
-from typing import TYPE_CHECKING
-
 from poetry.core.packages.constraints.base_constraint import BaseConstraint
 from poetry.core.packages.constraints.empty_constraint import EmptyConstraint
 
 
-if TYPE_CHECKING:
-    from poetry.core.packages.constraints import ConstraintTypes  # noqa
-
-
 class AnyConstraint(BaseConstraint):
-    def allows(self, other: "ConstraintTypes") -> bool:
+    def allows(self, other: "BaseConstraint") -> bool:
         return True
 
-    def allows_all(self, other: "ConstraintTypes") -> bool:
+    def allows_all(self, other: "BaseConstraint") -> bool:
         return True
 
-    def allows_any(self, other: "ConstraintTypes") -> bool:
+    def allows_any(self, other: "BaseConstraint") -> bool:
         return True
 
-    def difference(self, other: "ConstraintTypes") -> "ConstraintTypes":
+    def difference(self, other: "BaseConstraint") -> "BaseConstraint":
```

Comment:
Unless we can make the same change on the `BaseConstraint` then this I'd suggest this is somewhere in the region of confusing / not useful!

What you probably have in hand when performing `difference()` is a `BaseConstraint` anyway, so the specific type of the method on the subclass usually won't be interesting anyway.  It just needs to agree with the base type so that the typechecker is happy.

The proposed parameterisation is not true for all base types eg it doesn't make sense for an `AnyConstraint`.  (The current implementation on `AnyConstraint` _does_ satisfy this: but as previously noted, that's surely because it's a bogus implementation!)